From f5b51e70017bd39b0b7ab05c119e20d000e9c67b Mon Sep 17 00:00:00 2001
From: York Jasper Niebuhr <yjn@yjn-systems.com>
Date: Wed, 24 Jun 2026 13:53:30 +0200
Subject: [PATCH 2/3] Fieldlabel test

This patch adds a simple test to validate whether field labels are
inserted at the right offsets inside instructions.

Signed-off-by: York Jasper Niebuhr <yjn@yjn-systems.com>
---
 gas/testsuite/gas/i386/i386.exp            |   2 +
 gas/testsuite/gas/i386/x86-64-fieldlabel.d |   7 +
 gas/testsuite/gas/i386/x86-64-fieldlabel.s | 148 +++++++++++++++++++++
 3 files changed, 157 insertions(+)
 create mode 100644 gas/testsuite/gas/i386/x86-64-fieldlabel.d
 create mode 100644 gas/testsuite/gas/i386/x86-64-fieldlabel.s

diff --git a/gas/testsuite/gas/i386/i386.exp b/gas/testsuite/gas/i386/i386.exp
index f63f5df703e..e933458db97 100644
--- a/gas/testsuite/gas/i386/i386.exp
+++ b/gas/testsuite/gas/i386/i386.exp
@@ -903,4 +903,6 @@ if {[is_elf_format] || [istarget "*-*-vxworks*"]} then {
 	&& ![istarget "*-*-netbsd*"]} then {
 	run_dump_test "svr4"
     }
+
+	run_dump_test "x86-64-fieldlabel"
 }
diff --git a/gas/testsuite/gas/i386/x86-64-fieldlabel.d b/gas/testsuite/gas/i386/x86-64-fieldlabel.d
new file mode 100644
index 00000000000..91b32f064d8
--- /dev/null
+++ b/gas/testsuite/gas/i386/x86-64-fieldlabel.d
@@ -0,0 +1,7 @@
+#name: x86-64 fieldlabel
+#as: -L
+#objdump: -dr
+#target: x86_64-*-*
+
+.*: +file format .*
+#pass
diff --git a/gas/testsuite/gas/i386/x86-64-fieldlabel.s b/gas/testsuite/gas/i386/x86-64-fieldlabel.s
new file mode 100644
index 00000000000..598da20fb9d
--- /dev/null
+++ b/gas/testsuite/gas/i386/x86-64-fieldlabel.s
@@ -0,0 +1,148 @@
+# Test fieldlabel(EXPR, LABEL) on x86-64 instruction fields.
+#
+# Each test has a normal label at the start of the instruction and an
+# assembler-time assertion that the field label is at the expected offset
+# from that instruction start.  This avoids depending on absolute offsets
+# in .text.
+
+	.text
+	.globl foo
+foo:
+
+# Constant immediates.
+t_imm8:
+	mov $fieldlabel(0x12, .Limm8), %al
+	.if .Limm8 - t_imm8 != 1
+	.error "imm8 -> operand label is at unexpected offset"
+	.endif
+
+t_imm16:
+	mov $fieldlabel(0x1234, .Limm16), %ax
+	.if .Limm16 - t_imm16 != 2
+	.error "imm16 -> operand label is at unexpected offset"
+	.endif
+
+t_imm32:
+	mov $fieldlabel(0x12345678, .Limm32), %eax
+	.if .Limm32 - t_imm32 != 1
+	.error "imm32 -> operand label is at unexpected offset"
+	.endif
+
+t_imm64_32enc:
+	mov $fieldlabel(0x12345678, .Limm64_32enc), %rax
+	.if .Limm64_32enc - t_imm64_32enc != 3
+	.error "imm64_32enc -> operand label is at unexpected offset"
+	.endif
+
+t_imm64:
+	movabs $fieldlabel(0x1122334455667788, .Limm64), %rax
+	.if .Limm64 - t_imm64 != 2
+	.error "imm64 -> operand label is at unexpected offset"
+	.endif
+
+# Relocatable immediates.
+t_relimm32:
+	mov $fieldlabel(bar, .Lrelimm32), %eax
+	.if .Lrelimm32 - t_relimm32 != 1
+	.error "relimm32 -> operand label is at unexpected offset"
+	.endif
+
+t_relimm64:
+	mov $fieldlabel(bar, .Lrelimm64), %rax
+	.if .Lrelimm64 - t_relimm64 != 3
+	.error "relimm64 -> operand label is at unexpected offset"
+	.endif
+
+t_relabs64:
+	movabs $fieldlabel(bar, .Lrelabs64), %rax
+	.if .Lrelabs64 - t_relabs64 != 2
+	.error "relabs64 -> operand label is at unexpected offset"
+	.endif
+
+# RIP-relative memory operands.
+t_ripdisp1:
+	mov fieldlabel(bar, .Lripdisp1)(%rip), %rax
+	.if .Lripdisp1 - t_ripdisp1 != 3
+	.error "ripdisp1 -> operand label is at unexpected offset"
+	.endif
+
+t_ripdisp2:
+	lea fieldlabel(bar, .Lripdisp2)(%rip), %rax
+	.if .Lripdisp2 - t_ripdisp2 != 3
+	.error "ripdisp2 -> operand label is at unexpected offset"
+	.endif
+
+# Base register + displacement.
+t_disp8:
+	mov fieldlabel(16, .Ldisp8)(%rax), %rbx
+	.if .Ldisp8 - t_disp8 != 3
+	.error "disp8 -> operand label is at unexpected offset"
+	.endif
+
+t_disp32:
+	mov fieldlabel(0x12345678, .Ldisp32)(%rax), %rbx
+	.if .Ldisp32 - t_disp32 != 3
+	.error "disp32 -> operand label is at unexpected offset"
+	.endif
+
+# Indexed addressing.
+t_indexdisp:
+	mov fieldlabel(0x20, .Lindexdisp)(%rax,%rcx,4), %rbx
+	.if .Lindexdisp - t_indexdisp != 4
+	.error "indexdisp -> operand label is at unexpected offset"
+	.endif
+
+# Immediate and displacement in the same instruction.
+t_both:
+	addq $fieldlabel(5, .Laddimm), fieldlabel(32, .Ladddisp)(%rax)
+	.if .Ladddisp - t_both != 3
+	.error "both -> destination operand label is at unexpected offset"
+	.endif
+	.if .Laddimm - t_both != 4
+	.error "both -> source operand label is at unexpected offset"
+	.endif
+
+# Multiple independent immediates.
+t_multi1:
+	mov $fieldlabel(1, .Lmulti1), %eax
+	.if .Lmulti1 - t_multi1 != 1
+	.error "multi1 -> operand label is at unexpected offset"
+	.endif
+
+t_multi2:
+	mov $fieldlabel(2, .Lmulti2), %ebx
+	.if .Lmulti2 - t_multi2 != 1
+	.error "multi2 -> operand label is at unexpected offset"
+	.endif
+
+t_multi3:
+	mov $fieldlabel(3, .Lmulti3), %ecx
+	.if .Lmulti3 - t_multi3 != 1
+	.error "multi3 -> operand label is at unexpected offset"
+	.endif
+
+	ret
+
+	.data
+	.quad .Limm8
+	.quad .Limm16
+	.quad .Limm32
+	.quad .Limm64_32enc
+	.quad .Limm64
+	.quad .Lrelimm32
+	.quad .Lrelimm64
+	.quad .Lrelabs64
+	.quad .Lripdisp1
+	.quad .Lripdisp2
+	.quad .Ldisp8
+	.quad .Ldisp32
+	.quad .Lindexdisp
+	.quad .Laddimm
+	.quad .Ladddisp
+	.quad .Lmulti1
+	.quad .Lmulti2
+	.quad .Lmulti3
+
+	.globl bar
+bar:
+	.quad 0
-- 
2.43.0

